### ISSUE TITLE
Something wrong with the decoder result of Walkthrough example

### ISSUE THREAD

**Issue by lovecambi** [tensorflow/tensor2tensor#67]
I trained the model on two Tesla M60s, each of which is 8G. I did not modify any hyper-parameter. The loss seems not change after 50000 steps. 
```
INFO:tensorflow:Saving checkpoints for 54877 into /data/t2t/t2t_train/wmt_ende_tokens_32k/transformer-transformer_base/model.ckpt.
INFO:tensorflow:global_step/sec: 0.858429
INFO:tensorflow:loss = 1.96779, step = 54901 (116.492 sec)
INFO:tensorflow:global_step/sec: 0.876813
INFO:tensorflow:loss = 1.96174, step = 55001 (114.049 sec)
INFO:tensorflow:global_step/sec: 0.864947
INFO:tensorflow:loss = 1.98628, step = 55101 (115.614 sec)
INFO:tensorflow:global_step/sec: 0.860629
INFO:tensorflow:loss = 2.26156, step = 55201 (116.195 sec)
INFO:tensorflow:global_step/sec: 0.864128
INFO:tensorflow:loss = 1.98318, step = 55301 (115.723 sec)
INFO:tensorflow:Saving checkpoints for 55396 into /data/t2t/t2t_train/wmt_ende_tokens_32k/transformer-transformer_base/model.ckpt.
INFO:tensorflow:global_step/sec: 0.852946
INFO:tensorflow:loss = 2.30657, step = 55401 (117.241 sec)
INFO:tensorflow:global_step/sec: 0.870939
INFO:tensorflow:loss = 2.11571, step = 55501 (114.819 sec)
INFO:tensorflow:global_step/sec: 0.86979
INFO:tensorflow:loss = 1.99461, step = 55601 (114.970 sec)
INFO:tensorflow:global_step/sec: 0.86269
INFO:tensorflow:loss = 2.01496, step = 55701 (115.916 sec)
INFO:tensorflow:global_step/sec: 0.869183
INFO:tensorflow:loss = 1.98261, step = 55801 (115.051 sec)
INFO:tensorflow:global_step/sec: 0.862935
INFO:tensorflow:loss = 1.88075, step = 55901 (115.883 sec)
INFO:tensorflow:Saving checkpoints for 55915 into /data/t2t/t2t_train/wmt_ende_tokens_32k/transformer-transformer_base/model.ckpt.
INFO:tensorflow:global_step/sec: 0.855085
INFO:tensorflow:loss = 1.9415, step = 56001 (116.948 sec)
INFO:tensorflow:global_step/sec: 0.86353
INFO:tensorflow:loss = 2.26614, step = 56101 (115.804 sec)
INFO:tensorflow:global_step/sec: 0.871136
INFO:tensorflow:loss = 2.14308, step = 56201 (114.793 sec)
INFO:tensorflow:global_step/sec: 0.860752
INFO:tensorflow:loss = 1.96734, step = 56301 (116.178 sec)
INFO:tensorflow:global_step/sec: 0.871609
INFO:tensorflow:loss = 1.98928, step = 56401 (114.730 sec)
```

However, the decoder result does not make any sense. Anyone knows the reason?
```
INFO:tensorflow:Inference results INPUT: Goodbye world
INFO:tensorflow:Inference results OUTPUT: Esconnectentareaconnectentkannconnectent
INFO:tensorflow:Inference results INPUT: Hello world
INFO:tensorflow:Inference results OUTPUT: Esconnectentareaconnectentkannconnectent
```

**Comment by lukaszkaiser** [comment]
Can you just run the trainer with `--train_step=0 --eval_steps=10` and paste the results of the eval and example decodes? That'll help us see where the problem might be. Thanks!

**Comment by lovecambi** [comment]
@lukaszkaiser As your suggestion,
```
INFO:tensorflow:Saving dict for global step 64734: global_step = 64734, loss = 10.4774, 
metrics-wmt_ende_tokens_32k/accuracy = 6.01287e-05, 
metrics-wmt_ende_tokens_32k/accuracy_per_sequence = 0.0, 
metrics-wmt_ende_tokens_32k/accuracy_top5 = 0.0144008, 
metrics-wmt_ende_tokens_32k/bleu_score = 0.309825, 
metrics-wmt_ende_tokens_32k/neg_log_perplexity = -11.7884, 
metrics/accuracy = 6.01287e-05, 
metrics/accuracy_per_sequence = 0.0, 
metrics/accuracy_top5 = 0.0144008, 
metrics/bleu_score = 0.309825, 
metrics/neg_log_perplexity = -11.7884
```

**Comment by lukaszkaiser** [comment]
It looks like your training exploded and never got anywhere. Try `transformer_base_single_gpu` instead of `transformer_base` -- it has more warmup steps, so it shouldn't explode.

**Comment by vthorsteinsson** [comment]
...and a smaller `--batch_size` might also help, try 4096 or 2048. On my 8GB GeForce 1080 GTX I need to run `transformer_base_single_gpu` with batch size of max 4096.

**Comment by tobyyouup** [comment]
@lukaszkaiser   @lovecambi  Just re-generate the training sample.  I encounter some similar problems and fix by doing this.

**Comment by lovecambi** [comment]
Thanks, all. It does work.

**Comment by lukaszkaiser** [comment]
Thanks, I'm happy that it works now!

**Comment by viig99** [comment]
@lovecambi i am getting the same problem on a different dataset, can you please tell me what you tried that got it working, i ran transformer_base_single_gpu for 50000 iterations and i got similar results on inference, did running transformer_big_single_gpu for all 250000 iterations work for you ?

**Comment by lovecambi** [comment]
@viig99 I also set the ```--hparams='batch_size=2048'```. Can you try it?

**Comment by viig99** [comment]
I am using the transformer_big_single_gpu with batch_size 1024 and this is what my loss and accuracy looks like
![Accuracy & Loss](https://user-images.githubusercontent.com/990443/28002306-891f9b76-6551-11e7-9cb4-0f5417a46234.png)

the loss seems to be fluctuating in the region of 6.70 - 7.50, and the accuracy doesn't seem to be improving much, it has run for around 70000 iterations till now, is this expected and should i run it till 250000 ? thanks for the help guys.

**Comment by lukaszkaiser** [comment]
The graph you pasted is how it always looks when the training explodes in the first phase. It's not learning, waiting longer won't help. You need to (a) lower your learning rate and maybe (b) increase warmup steps. Try a learning rate that's 3x or 5x smaller and let me know, you should see the difference by 20k steps.

**Comment by viig99** [comment]
@lukaszkaiser thank you for the quick reply, i tried a 10x slower learning rate and bumped up the warmup steps to 20k this is the graph that i got around 25k steps,
![screenshot from 2017-07-11 09-40-21](https://user-images.githubusercontent.com/990443/28050720-0df1ad5a-661d-11e7-8b5e-a5130887f401.png)

i am still getting some fluctuations is this expected and does it smooth over longer steps, or would it better to lower the learning rate further until i get a smoother curve ?

**Comment by lukaszkaiser** [comment]
It looks like you're fine for now. Let your model train a bit now (another 20-40k steps) and you'll see how it goes. It might be training slow now with the low learning rate, but you'll have a baseline and you can try higher ones later. Let me know how your loss looks later in the training.

**Comment by viig99** [comment]
@lukaszkaiser as per you suggestion after another 60k steps the accuracy and neg log perplexity are improving slowly, loss function seems to be fluctuating a lot but but very slowly going down.
![screenshot from 2017-07-11 15-57-51](https://user-images.githubusercontent.com/990443/28064344-dc42be68-6651-11e7-9f8e-59e339f4910b.png)

**Comment by lukaszkaiser** [comment]
Ok, so maybe it's time to try with a 3x larger learning rate this time to speed things up.

**Comment by viig99** [comment]
@lukaszkaiser i tried the problem with a 3x larger learning rate and the accuracy started decreasing after around 40k iterations, also i let the above learning learning run for 250k iterations and my accuracy was around 0.14, on inference for all examples i got just one symbol in the end sequence following by eos and pad's, any idea what could be going here ?

**Comment by lukaszkaiser** [comment]
Looks like your model is still not training too well. I guess the easiest thing would be to add your dataset (or a small version) to tensor2tensor, so we could also take a look.
